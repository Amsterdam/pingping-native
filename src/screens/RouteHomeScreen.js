import React from 'react';

import { useLazyQuery } from '@apollo/client';
import PropTypes from 'prop-types';
import {
	Animated,
	RefreshControl,
	StyleSheet,
	View,
	ScrollView,
} from 'react-native';
import { View as AnimatableView } from 'react-native-animatable';

import { testIDs } from '../../e2e/modulesTestIDs';
import GET_ROUTES from '../apollo/Query/getRoutes';
import ContentLayout from '../components/layout/ContentLayout';
import RouteCard from '../components/route/RouteCard';
import EmptyContentNotifier from '../components/shared/EmptyContentNotifier';
import ErrorComponent from '../components/shared/ErrorComponent';
import FocusAwareStatusBar from '../components/shared/FocusAwareStatusBar';
import CardSkeleton from '../components/skeleton/CardSkeleton';
import Title from '../components/typography/Title';
import theme from '../config/theme';

const HEADER_HEIGHT = 200;

const RouteHomeScreen = ({ navigation }) => {
	React.useEffect(() => {
		const unsubscribe = navigation.addListener(
			'focus',
			() => {
				getRoutes();
			},
		);
		return unsubscribe;
	}, [navigation, getRoutes]);

	const scrollY = new Animated.Value(0);
	const translateY = scrollY.interpolate({
		inputRange: [0, HEADER_HEIGHT],
		outputRange: [0, -HEADER_HEIGHT],
	});

	const [getRoutes, routes] = useLazyQuery(
		GET_ROUTES,
		{
			fetchPolicy: 'cache-and-network',
		},
	);
	const [
		refreshing,
		setRefreshing,
	] = React.useState(false);

	if (routes.error) {
		return (
			<ErrorComponent
				functionToRetry={routes.refetch}
				somethingWentWrong
				onPress={routes.refetch}
				deafultLabelOverRide="Probeer Opnieuw"
			/>
		);
	}

	const onRefresh = () => {
		setRefreshing(true);
		routes.refetch();
		setRefreshing(false);
	};

	function compareProgress(a, b) {
		return b.progress - a.progress;
	}

	const renderRoutes = () => {
		const {
			availableRoutes,
			currentRoutes,
			//   archivedRoutes,
		} = routes.data.getRoutes;

		const suggestedRoutes = [];
		const otherRoutes = [];
		const mergedRoutes = [
			...availableRoutes,
			...currentRoutes,
		];
		mergedRoutes.forEach(route => {
			if (route.isSuggested) {
				return suggestedRoutes.push(route);
			}
			return otherRoutes.push(route);
		});

		suggestedRoutes.sort(compareProgress);
		otherRoutes.sort(compareProgress);

		return (
			<AnimatableView animation="fadeIn">
				{suggestedRoutes.length > 0 && (
					<React.Fragment>
						<Title
							style={styles.title}
							variant="h2"
							align="left"
						>
							Aanbevolen
						</Title>
						{suggestedRoutes.map(route => (
							<RouteCard
								navigation={navigation}
								route={route}
								key={route.routeId}
							/>
						))}
					</React.Fragment>
				)}

				<React.Fragment>
					<Title
						variant="h3"
						align="left"
						style={
							suggestedRoutes.length > 0
								? styles.subTitle
								: styles.title
						}
					>
						Andere life events
					</Title>

					<EmptyContentNotifier text="In de toekomst krijg je een notificatie wanneer een nieuwe route beschikbaar is." />

					{otherRoutes.map(route => (
						<RouteCard
							navigation={navigation}
							route={route}
							key={route.routeId}
						/>
					))}
				</React.Fragment>
			</AnimatableView>
		);
	};

	return (
		<View
			style={styles.container}
			testID={testIDs.ROUTES.SCREEN}
		>
			<View
				style={styles.underLayer}
				testID={testIDs.ROUTES.ANIMATED_VIEW}
			/>
			<FocusAwareStatusBar
				backgroundColor={theme.colors.headerColor}
				barStyle="light-content"
			/>
			<Animated.View
				style={[
					styles.header,
					{
						transform: [
							{ translateY: translateY },
						],
					},
				]}
			/>

			<ScrollView
				onScroll={e => {
					scrollY.setValue(
						e.nativeEvent.contentOffset.y,
					);
				}}
				scrollEventThrottle={16}
				contentContainerStyle={styles.content}
				refreshControl={
					<RefreshControl
						refreshing={refreshing}
						onRefresh={onRefresh}
						tintColor={theme.colors.primary}
						style={{
							backgroundColor:
								theme.colors.headerColor,
						}}
					/>
				}
			>
				<ContentLayout>
					{!routes.data && (
						<AnimatableView animation="fadeIn">
							<CardSkeleton />
							<CardSkeleton />
						</AnimatableView>
					)}
					{routes.data && renderRoutes()}
				</ContentLayout>
			</ScrollView>
		</View>
	);
};

const styles = StyleSheet.create({
	container: {
		backgroundColor: theme.colors.headerColor,
		position: 'relative',
	},
	header: {
		flexDirection: 'column',
		backgroundColor: theme.colors.headerColor,
		height: HEADER_HEIGHT,
		left: 0,
		top: 0,
		right: 0,
		position: 'absolute',
	},
	title: {
		color: theme.colors.white,
	},
	subTitle: {
		color: theme.colors.primary,
		marginVertical: theme.spacing.xs,
	},
	content: {
		top: 25,
		paddingBottom: theme.spacing.multiplier(15),
	},
	underLayer: {
		position: 'absolute',
		flex: 1,
		zIndex: -5,
		elevation: 0,
		backgroundColor: theme.colors.almostNotBlue,
		top: 100, // replace this with a percentage of the screenheight to be responsive
		bottom: 0,
		left: 0,
		right: 0,
	},
});

RouteHomeScreen.propTypes = {
	navigation: PropTypes.object.isRequired,
};

export default RouteHomeScreen;
