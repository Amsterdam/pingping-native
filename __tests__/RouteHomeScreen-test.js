import React from 'react';

import { MockedProvider } from '@apollo/client/testing';
import { NativeBaseProvider } from 'native-base';
import renderer from 'react-test-renderer';

import RouteHomeScreen from '../src/screens/RouteHomeScreen';

jest.useFakeTimers();
const navigation = {
	navigate: jest.fn(),
	addListener: jest.fn(),
};

test('renders correctly', () => {
	const tree = renderer
		.create(
			<NativeBaseProvider>
				<MockedProvider mocks={[]}>
					<RouteHomeScreen
						navigation={navigation}
					/>
				</MockedProvider>
			</NativeBaseProvider>,
		)
		.toJSON();
	expect(tree).toMatchSnapshot();
});
