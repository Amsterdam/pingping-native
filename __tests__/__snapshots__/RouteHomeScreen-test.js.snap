// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "height": 1334,
      },
      Object {
        "backgroundColor": "#0d2036",
      },
    ]
  }
  testID="ROUTES_SCREEN"
>
  <View
    style={
      Object {
        "backgroundColor": "#f7fbff",
        "bottom": 0,
        "elevation": 0,
        "flex": 1,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 100,
        "zIndex": -5,
      }
    }
    testID="ANIMATED_VIEW"
  />
  <View
    collapsable={false}
    nativeID="animatedComponent"
    noShadow={true}
    style={
      Object {
        "backgroundColor": "#0d2036",
        "flexDirection": "column",
        "height": 200,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
    transparent={true}
  />
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "backgroundColor": "transparent",
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      automaticallyAdjustContentInsets={false}
      contentContainerStyle={
        Array [
          Object {
            "padding": undefined,
          },
          Object {
            "paddingBottom": 75,
            "position": "absolute",
            "top": 25,
          },
        ]
      }
      contentInset={
        Object {
          "bottom": 0,
        }
      }
      enableAutomaticScroll={true}
      enableOnAndroid={false}
      enableResetScrollToCoords={true}
      extraHeight={75}
      extraScrollHeight={0}
      getScrollResponder={[Function]}
      handleOnScroll={[Function]}
      keyboardDismissMode="interactive"
      keyboardOpeningTime={250}
      keyboardShouldPersistTaps="handled"
      keyboardSpace={0}
      onScroll={[Function]}
      refreshControl={
        <RefreshControlMock
          onRefresh={[Function]}
          refreshing={false}
          style={
            Object {
              "backgroundColor": "#0d2036",
            }
          }
          tintColor="#fb9f00"
        />
      }
      resetKeyboardSpace={[Function]}
      resetScrollToCoords={
        Object {
          "x": 0,
          "y": 0,
        }
      }
      scrollEventThrottle={16}
      scrollForExtraHeightOnAndroid={[Function]}
      scrollIntoView={[Function]}
      scrollToEnd={[Function]}
      scrollToFocusedInput={[Function]}
      scrollToPosition={[Function]}
      showsVerticalScrollIndicator={true}
      style={
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
        }
      }
      update={[Function]}
      viewIsInsideTabBar={false}
    >
      <RCTRefreshControl />
      <View>
        <View
          style={
            Object {
              "flex": 1,
              "padding": 25,
              "position": "relative",
              "width": 750,
            }
          }
        >
          <View
            forwardedRef={[Function]}
            style={
              Object {
                "opacity": 0,
              }
            }
          >
            <View
              onLayout={[Function]}
            >
              <View
                style={
                  Object {
                    "position": "relative",
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderRadius": 4,
                        "height": 30,
                        "marginVertical": 10,
                        "width": "90%",
                      },
                      Object {
                        "backgroundColor": "#E1E9EE",
                        "overflow": "hidden",
                      },
                    ]
                  }
                />
              </View>
              <View
                style={
                  Array [
                    Object {
                      "alignSelf": "stretch",
                      "backgroundColor": "#fff",
                      "borderRadius": 5,
                      "marginVertical": 10,
                    },
                    Object {
                      "elevation": 5,
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 2,
                        "width": 0,
                      },
                      "shadowOpacity": 0.25,
                      "shadowRadius": 3.84,
                    },
                  ]
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "position": "relative",
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "height": 125,
                            "width": "100%",
                          },
                          Object {
                            "backgroundColor": "#E1E9EE",
                            "overflow": "hidden",
                          },
                        ]
                      }
                    />
                  </View>
                  <View
                    style={
                      Object {
                        "marginTop": 10,
                        "padding": 20,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "borderRadius": 4,
                              "height": 20,
                              "width": "90%",
                            },
                            Object {
                              "backgroundColor": "#E1E9EE",
                              "overflow": "hidden",
                            },
                          ]
                        }
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "borderRadius": 4,
                              "height": 20,
                              "marginTop": 6,
                              "width": "80%",
                            },
                            Object {
                              "backgroundColor": "#E1E9EE",
                              "overflow": "hidden",
                            },
                          ]
                        }
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "borderRadius": 4,
                              "height": 20,
                              "marginTop": 6,
                              "width": 80,
                            },
                            Object {
                              "backgroundColor": "#E1E9EE",
                              "overflow": "hidden",
                            },
                          ]
                        }
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignSelf": "flex-end",
                              "borderRadius": 4,
                              "height": 10,
                              "marginTop": 6,
                              "width": 50,
                            },
                            Object {
                              "backgroundColor": "#E1E9EE",
                              "overflow": "hidden",
                            },
                          ]
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
            <View
              onLayout={[Function]}
            >
              <View
                style={
                  Object {
                    "position": "relative",
                  }
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "borderRadius": 4,
                        "height": 30,
                        "marginVertical": 10,
                        "width": "90%",
                      },
                      Object {
                        "backgroundColor": "#E1E9EE",
                        "overflow": "hidden",
                      },
                    ]
                  }
                />
              </View>
              <View
                style={
                  Array [
                    Object {
                      "alignSelf": "stretch",
                      "backgroundColor": "#fff",
                      "borderRadius": 5,
                      "marginVertical": 10,
                    },
                    Object {
                      "elevation": 5,
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 2,
                        "width": 0,
                      },
                      "shadowOpacity": 0.25,
                      "shadowRadius": 3.84,
                    },
                  ]
                }
              >
                <View>
                  <View
                    style={
                      Object {
                        "position": "relative",
                      }
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "height": 125,
                            "width": "100%",
                          },
                          Object {
                            "backgroundColor": "#E1E9EE",
                            "overflow": "hidden",
                          },
                        ]
                      }
                    />
                  </View>
                  <View
                    style={
                      Object {
                        "marginTop": 10,
                        "padding": 20,
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "borderRadius": 4,
                              "height": 20,
                              "width": "90%",
                            },
                            Object {
                              "backgroundColor": "#E1E9EE",
                              "overflow": "hidden",
                            },
                          ]
                        }
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "borderRadius": 4,
                              "height": 20,
                              "marginTop": 6,
                              "width": "80%",
                            },
                            Object {
                              "backgroundColor": "#E1E9EE",
                              "overflow": "hidden",
                            },
                          ]
                        }
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "borderRadius": 4,
                              "height": 20,
                              "marginTop": 6,
                              "width": 80,
                            },
                            Object {
                              "backgroundColor": "#E1E9EE",
                              "overflow": "hidden",
                            },
                          ]
                        }
                      />
                    </View>
                    <View
                      style={
                        Object {
                          "position": "relative",
                        }
                      }
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "alignSelf": "flex-end",
                              "borderRadius": 4,
                              "height": 10,
                              "marginTop": 6,
                              "width": 50,
                            },
                            Object {
                              "backgroundColor": "#E1E9EE",
                              "overflow": "hidden",
                            },
                          ]
                        }
                      />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </RCTSafeAreaView>
</View>
`;
